src/Sulu/Bundle/WebsiteBundle/EventListener/RedirectExceptionListener.php

```diff
@@ -0,0 +1,241 @@
+        RequestAnalyzerInterface $requestAnalyzer,
+        DefaultLocaleProviderInterface $defaultLocaleProvider,
+        ReplacerInterface $urlReplacer
+    ) {
+        $this->router = $router;
+        $this->requestAnalyzer = $requestAnalyzer;
+        $this->defaultLocaleProvider = $defaultLocaleProvider;
+        $this->urlReplacer = $urlReplacer;
+    }
+
+    /**
+     * Redirect trailing slashes or ".html".
+     *
+     * @param GetResponseForExceptionEvent $event
+     */
+    public function redirectTrailingSlashOrHtml(GetResponseForExceptionEvent $event)
+    {
+        $request = $event->getRequest();
+        $route = $this->getRoute($request);
+        if ($route === $request->getPathInfo() || !$this->matchRoute($route, $request->getSchemeAndHttpHost())) {
+            return;
+        }
+
+        $event->setResponse(new RedirectResponse($route, 301));
+    }
+
+    /**
+     * Redirect partial and redirect matches.
+     *
+     * @param GetResponseForExceptionEvent $event
+     */
+    public function redirectPartialMatch(GetResponseForExceptionEvent $event)
+    {
+        $request = $event->getRequest();
+
+        /** @var RequestAttributes $attributes */
+        $attributes = $event->getRequest()->attributes->get('_sulu', new RequestAttributes());
+
+        $types = [RequestAnalyzerInterface::MATCH_TYPE_REDIRECT, RequestAnalyzerInterface::MATCH_TYPE_PARTIAL];
+        $matchType = $attributes->getAttribute('matchType');
+        if (!in_array($matchType, $types)) {
+            return;
+        }
+
+        $localization = $this->defaultLocaleProvider->getDefaultLocale();
+
+        $redirect = $attributes->getAttribute('redirect');
+        $redirect = $this->urlReplacer->replaceCountry($redirect, $localization->getCountry());
+        $redirect = $this->urlReplacer->replaceLanguage($redirect, $localization->getLanguage());
+        $redirect = $this->urlReplacer->replaceLocalization($redirect, $localization->getLocale(Localization::DASH));
+
+        $route = $this->resolveRedirectUrl(
+            $redirect,
+            $request->getUri(),
+            $attributes->getAttribute('resourceLocatorPrefix')
+        );
+
+        if (!$this->matchRoute($route, $request->getSchemeAndHttpHost())) {
+            return;
+        }
+
+        $event->setResponse(new RedirectResponse($route, 301));
+    }
+
+    /**
+     * Returns true if given route exists.
+     *
+     * @param string $route
+     * @param string $domain
+     *
+     * @return bool
+     */
+    private function matchRoute($route, $domain)
+    {
+        return $this->matchUrl($domain . $route);
+    }
+
+    /**
+     * Returns true if given url exists.
+     *
+     * @param string $url
+     *
+     * @return bool
+     */
+    private function matchUrl($url)
+    {
+        $request = Request::create($url);
+        $this->requestAnalyzer->analyze($request);
+
+        try {
+            return $this->router->matchRequest($request) !== null;
```

Comment:
we need the `RequestAttributes` which will be added to this request a few lines earlier. this cannt be done in the match method.